File: internal/api/middleware.go
package api

import (
	"log/slog"
	"net/http"
	"strings"
	"time"
)

func LoggingMiddleware(logger *slog.Logger) func(http.Handler) http.Handler {
	if logger == nil {
		logger = slog.Default()
	}
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			start := time.Now()
			next.ServeHTTP(w, r)
			logger.Info("http",
				"method", r.Method,
				"path", r.URL.Path,
				"remote", r.RemoteAddr,
				"ua", r.UserAgent(),
				"dur", time.Since(start).String(),
			)
		})
	}
}

func RecoverMiddleware(logger *slog.Logger) func(http.Handler) http.Handler {
	if logger == nil {
		logger = slog.Default()
	}
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			defer func() {
				if rec := recover(); rec != nil {
					logger.Error("panic", "err", rec)
					w.WriteHeader(http.StatusInternalServerError)
				}
			}()
			next.ServeHTTP(w, r)
		})
	}
}

func AuthMiddleware(token string) func(http.Handler) http.Handler {
	token = strings.TrimSpace(token)
	if token == "" {
		return func(h http.Handler) http.Handler { return h }
	}
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			auth := r.Header.Get("Authorization")
			const prefix = "Bearer "
			if !strings.HasPrefix(auth, prefix) || strings.TrimSpace(strings.TrimPrefix(auth, prefix)) != token {
				w.WriteHeader(http.StatusUnauthorized)
				return
			}
			next.ServeHTTP(w, r)
		})
	}
}

// CORS is intentionally minimal. For remote access, prefer an authenticated tunnel.
func CORSMiddleware() func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.Header().Set("Access-Control-Allow-Origin", "*")
			w.Header().Set("Access-Control-Allow-Headers", "Authorization, Content-Type")
			w.Header().Set("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
			if r.Method == http.MethodOptions {
				w.WriteHeader(http.StatusNoContent)
				return
			}
			next.ServeHTTP(w, r)
		})
	}
}

File: internal/api/middleware.ts
export type Handler = (req: Request) => Promise<Response>;
export type Middleware = (req: Request, next: Handler) => Promise<Response>;

export function loggingMiddleware(logger: Console = console): Middleware {
  return async (req, next) => {
    const start = Date.now();
    const res = await next(req);
    const duration = Date.now() - start;
    logger.info("http", {
      method: req.method,
      path: new URL(req.url).pathname,
      ua: req.headers.get("user-agent") ?? "",
      duration_ms: duration,
    });
    return res;
  };
}

export function recoverMiddleware(logger: Console = console): Middleware {
  return async (req, next) => {
    try {
      return await next(req);
    } catch (err) {
      logger.error("panic", err);
      return new Response(null, { status: 500 });
    }
  };
}

export function authMiddleware(token: string): Middleware {
  const trimmed = token.trim();
  if (!trimmed) {
    return async (_, next) => next(_);
  }
  return async (req, next) => {
    const auth = req.headers.get("authorization") ?? "";
    const prefix = "Bearer ";
    if (!auth.startsWith(prefix) || auth.slice(prefix.length).trim() !== trimmed) {
      return new Response(null, { status: 401 });
    }
    return next(req);
  };
}

export function corsMiddleware(): Middleware {
  return async (req, next) => {
    if (req.method === "OPTIONS") {
      return new Response(null, {
        status: 204,
        headers: {
          "Access-Control-Allow-Origin": "*",
          "Access-Control-Allow-Headers": "Authorization, Content-Type",
          "Access-Control-Allow-Methods": "GET, POST, OPTIONS",
        },
      });
    }
    const res = await next(req);
    const headers = new Headers(res.headers);
    headers.set("Access-Control-Allow-Origin", "*");
    headers.set("Access-Control-Allow-Headers", "Authorization, Content-Type");
    headers.set("Access-Control-Allow-Methods", "GET, POST, OPTIONS");
    return new Response(res.body, { status: res.status, statusText: res.statusText, headers });
  };
}

export function applyMiddleware(handler: Handler, middlewares: Middleware[]): Handler {
  return middlewares.reduceRight(
    (next, middleware) => {
      return (req) => middleware(req, next);
    },
    handler,
  );
}
